File: src/main/java/net/dean/jraw/http/Credentials.java
package net.dean.jraw.http;

/**
 * This class is responsible for handling any kind of Reddit user and OAuth2 credentials
 */
public final class Credentials {
    private final AuthenticationMethod authenticationMethod;
    private final String username;
    private final String password;
    private final String clientId;
    private final String clientSecret;

    private Credentials(AuthenticationMethod authenticationMethod, String username, String password, String clientId,
                       String clientSecret) {
        this.authenticationMethod = authenticationMethod;
        this.username = username;
        this.password = password;
        this.clientId = clientId;
        this.clientSecret = clientSecret;
    }

    /**
     * Gets how these Credentials are to be used
     * @return The AuthenticationMethod
     */
    public AuthenticationMethod getAuthenticationMethod() {
        return authenticationMethod;
    }

    public String getUsername() {
        return username;
    }

    public String getPassword() {
        return password;
    }

    /**
     * Gets the OAuth2 app's ID. Applicable for all OAuth app types.
     * @return The app's ID
     */
    public String getClientId() {
        return clientId;
    }

    /**
     * Gets the OAuth2 app's secret key. Applicable for all OAuth app types, however, for 'installed app' app types, the
     * secret will always be an empty string.
     *
     * @return The secret value for the application
     */
    public String getClientSecret() {
        return clientSecret;
    }

    /**
     * Creates new new Credentials object for "normal" (non-OAuth2) authentication
     * @param username The username
     * @param password The password
     * @return A new Credentials
     */
    public static Credentials standard(String username, String password) {
        return new Credentials(AuthenticationMethod.STANDARD, username, password, null, null);
    }

    /**
     * Creates a new Credentials object for a Reddit app whose type is 'installed app'
     * @param username The username
     * @param password The password
     * @param clientId The publicly-available app ID
     * @return A new Credentials
     */
    public static Credentials oauth2App(String username, String password, String clientId) {
        // Use an empty string as the client secret because "The 'password' for non-confidential clients
        // (installed apps) is an empty string". See https://github.com/reddit/reddit/wiki/OAuth2#token-retrieval
        return new Credentials(AuthenticationMethod.APP, username, password, clientId, "");
    }

    /**
     * Creates a new Credentails object for a Reddit app whose type is 'web app'
     * @param username The username
     * @param password The password
     * @param clientId The publicly-available app ID
     * @param clientSecret The secret value for the application
     * @return A new Credentials
     */
    public static Credentials oauth2Script(String username, String password, String clientId, String clientSecret) {
        return new Credentials(AuthenticationMethod.SCRIPT, username, password, clientId, clientSecret);
    }

    /**
     * Creates a new Credentails object for a Reddit app whose type is 'web app'
     * @param username The username
     * @param password The password
     * @param clientId The publicly-available app ID
     * @param clientSecret The secret value for the application
     * @return A new Credentials
     */
    public static Credentials oauth2Webapp(String username, String password, String clientId, String clientSecret) {
        return new Credentials(AuthenticationMethod.WEBAPP, username, password, clientId, clientSecret);
    }
}

File: lib/src/main/kotlin/net/dean/jraw/http/oauth/Credentials.kt
package net.dean.jraw.http.oauth

import java.util.*

class Credentials private constructor(
    /** How these credential were meant to be used */
    val authMethod: AuthMethod,

    /** Reddit username. Only non-null for script Credentials */
    val username: String? = null,

    /** Reddit password. Only non-null for script Credentials */
    val password: String? = null,

    /** OAuth2 app client ID. */
    val clientId: String,

    /** OAuth2 app client secret. This is an empty string for app or userless app Credentials. */
    val clientSecret: String,

    /** A unique, per-device ID */
    val deviceId: UUID? = null,

    /**
     * The URL that users will be redirected to when authorizing your application. Only non-null for installed or web
     * apps.
     */
    val redirectUrl: String? = null
) {
    companion object {
        @JvmStatic fun script(username: String, password: String, clientId: String, clientSecret: String) =
            Credentials(AuthMethod.SCRIPT,
                username = username,
                password = password,
                clientId = clientId,
                clientSecret = clientSecret)

        @JvmStatic fun installedApp(clientId: String, redirectUrl: String) =
            Credentials(AuthMethod.APP,
                clientId = clientId,
                clientSecret = "",
                redirectUrl = redirectUrl)

        @JvmStatic fun webapp(clientId: String, clientSecret: String, redirectUrl: String) =
            Credentials(AuthMethod.WEBAPP,
                clientId = clientId,
                clientSecret = clientSecret,
                redirectUrl = redirectUrl)

        @JvmStatic fun userless(clientId: String, clientSecret: String, deviceId: UUID) =
            Credentials(AuthMethod.USERLESS,
                clientId = clientId,
                clientSecret = clientSecret,
                deviceId = deviceId)

        @JvmStatic fun userlessApp(clientId: String, deviceId: UUID) =
            Credentials(AuthMethod.USERLESS_APP,
                clientId = clientId,
                clientSecret = "",
                deviceId = deviceId)

    }
}
